### DISCUSSION TITLE
Trouble with linking libsndfile in other cmake project

### DISCUSSION THREAD

**Discussion by DBraun** [libsndfile/libsndfile#800] [General]
On MacOS (specifically github actions), I'm trying to build libsndfile and then link against it in another project configured with CMake.

First I build libsndfile which is in a folder thirdparty/libsndfile.
```bash
brew update
brew install autoconf autogen automake flac libogg libtool libvorbis opus mpg123 pkg-config
mkdir thirdparty/libsndfile/build
cd thirdparty/libsndfile/build
cmake .. -G "Unix Makefiles"
make & make install
```
I've also tried it with `cmake --build . --config Release` at the end. Either way, building works.

Then I try to link against libsndfile. The CMake contains
```cmake
list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/thirdparty/libsndfile/cmake)
find_package(SndFile REQUIRED)
```
This fails because the find_package goes to https://github.com/libsndfile/libsndfile/blob/9349a566e298ef6d6bed6c275dc1f5502bb2e028/cmake/SndFileConfig.cmake.in#L19 which evaluates to `/Users/runner/work/My-Project/My-Project/thirdparty/libsndfile/build/SndFileTargets.cmake` which doesn't exist. I don't know why this file doesn't exist. It doesn't show up on Github actions if I do `ls -r *` after building libsndfile but files such as SndFileConfig.cmake do. On Windows, if I build libsndfile, I do get `C:\My-Project\thirdparty\libsndfile\build\CMakeFiles\Export\cmake\SndFileTargets.cmake`

Is the problem with creating/locating this SndFileTargets.cmake?

**Comment by evpobr** [comment]
Hi @DBraun . I was planning to add this feature, but at some point I forgot about it.

I will make PR later, the solution is:

```cmake
export (TARGETS sndfile NAMESPACE SndFile:: FILE ${PROJECT_BINARY_DIR}/SndFileTargets.cmake)
```

after

```cmake
install (TARGETS sndfile
        ...
```

If you check it out, I will be very grateful.

**Comment by DBraun** [comment]
Thanks I'm trying that out. Now the situation seems to be that https://github.com/libsndfile/libsndfile/blob/9349a566e298ef6d6bed6c275dc1f5502bb2e028/cmake/SndFileConfig.cmake.in#L21 is evaluating to `/Users/runner/work/My-Project/My-Project/include` but I think it needs to evaluate to `/Users/runner/work/My-Project/My-Project/thirdparty/libsndfile/build/include`

**Reply by DBraun** [reply]
I have a similar issue with the latest libsndfile. I think that this [`set_and_check`](https://github.com/libsndfile/libsndfile/blob/eff5f0a640ef2a21ac77b42aa22d42b30fd46c97/cmake/SndFileConfig.cmake.in#L27) line is evaluating to 1: `/__w/MyProject/MyProject/libsndfile/CMakeBuild/include` but I think it should evaluate to 2: `/__w/MyProject/MyProject/libsndfile/include`.

I use GitHub actions on ubuntu-latest to build libsndfile:
```bash
cd libsndfile
mkdir CMakeBuild && cd CMakeBuild
cmake .. -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DENABLE_EXTERNAL_LIBS=off -DBUILD_EXAMPLES=OFF
cmake --build . --config Release
```

Then in my main project, I use cmake with `-DSndFile_DIR="$PWD/libsndfile/CMakeBuild" -DCMAKE_PREFIX_PATH="$PWD/libsndfile/cmake"`

My `CMakeLists.txt` contains `find_package(SndFile REQUIRED)` which succeeds because that `CMakeBuild` directory contains `SndFileConfig.cmake`. However, I end up on the `set_and_check` line linked above.
```
CMake Error at /__w/MyProject/MyProject/libsndfile/CMakeBuild/SndFileConfig.cmake:19 (message):
  File or directory /__w/MyProject/MyProject/libsndfile/CMakeBuild/include referenced
  by variable SndFile_INCLUDE_DIR does not exist !
```

Was I supposed to use `CMAKE_INSTALL_PREFIX` when building libsndfile?

**Comment by evpobr** [comment]
Check #802 please.

**Comment by DBraun** [comment]
Almost there! These are the errors building My-Project
```
CMake Error at CMakeLists.txt:70 (add_library):
  Target "My-Project" links to target "MPG123::libmpg123" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?

CMake Error at CMakeLists.txt:70 (add_library):
  Target "My-Project" links to target "Lame::Lame" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
```

In this section https://github.com/libsndfile/libsndfile/blob/f44c8dbe1d30988801104c95589f1eecae340134/cmake/SndFileConfig.cmake.in#L13-L16
Should we add
```
find_dependency(Lame)
find_dependency(MPG123::libmpg123)
```
?

I'll see if I can work that into my own CMakeLists.txt rather than modifying libsndfile.

**Reply by evpobr** [reply]
Oh, it is bug. 

It should be:

```cmake
find_dependency(Lame)
find_dependency(MPG123)
```

**Reply by evpobr** [reply]
Check update please.

**Comment by DBraun** [comment]
All good now! Thanks 👍🏻

**Reply by evpobr** [reply]
Great!
